File: palettier-core/src/test/java/be/asmolabs/palettier/core/service/PaletteMixServiceTest.java
package be.asmolabs.palettier.core.service;

import static org.assertj.core.api.Assertions.assertThat;

import be.asmolabs.palettier.core.color.Rgb;
import be.asmolabs.palettier.core.domain.DryingClass;
import be.asmolabs.palettier.core.domain.PaletteMix;
import be.asmolabs.palettier.core.domain.Ventilation;
import be.asmolabs.palettier.core.service.DryingModels.Workshop;
import be.asmolabs.palettier.core.service.PaletteMixService.MixState;
import java.time.Duration;
import java.time.Instant;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
@Transactional
class PaletteMixServiceTest {

    private static final Instant NOW = Instant.parse("2026-03-01T10:00:00Z");

    @Autowired
    private PaletteMixService mixes;

    private PaletteMix gris(DryingClass dryingClass, Workshop workshop) {
        return mixes.record("Gris rompu des ombres", Rgb.ofHex("#6B6259"),
                "2 parts de terre d'ombre + 1 part de blanc", dryingClass, workshop, NOW);
    }

    @Test
    @DisplayName("un melange frais est ouvert, et le restera quelques heures")
    void afreshMixIsOpen() {
        PaletteMix mix = gris(DryingClass.MEDIUM, Workshop.standard());

        MixState state = mixes.state(mix, NOW.plus(Duration.ofMinutes(30)));

        assertThat(state.isOpen()).isTrue();
        assertThat(state.isSpent()).isFalse();
        assertThat(state.workable()).isPositive();
    }

    @Test
    @DisplayName("passe le temps ouvert il ne se travaille plus, et plus tard il a pris")
    void amixCloses() {
        PaletteMix mix = gris(DryingClass.FAST, Workshop.standard());

        // Tres au-dela de tout jalon : il a pris, il n'a plus rien a faire la.
        MixState spent = mixes.state(mix, NOW.plus(Duration.ofDays(20)));
        assertThat(spent.isOpen()).isFalse();
        assertThat(spent.isSpent()).isTrue();
    }

    @Test
    @DisplayName("le temps ouvert depend de l'atelier du moment ou l'on a melange")
    void theworkshopOfTheMomentDecides() {
        PaletteMix cold = mixes.record("Froid", Rgb.ofHex("#6B6259"), "", DryingClass.MEDIUM,
                new Workshop(10, 80, Ventilation.CONFINED), NOW);
        PaletteMix warm = mixes.record("Chaud", Rgb.ofHex("#6B6259"), "", DryingClass.MEDIUM,
                new Workshop(28, 30, Ventilation.GOOD), NOW);

        Instant later = NOW.plus(Duration.ofHours(1));
        assertThat(mixes.state(cold, later).workable())
                .as("un atelier froid et humide garde le melange ouvert plus longtemps")
                .isGreaterThan(mixes.state(warm, later).workable());
    }

    @Test
    @DisplayName("une pate sur la palette tient plus longtemps que la meme couche sur la piece")
    void apuddleOutlivesAfilm() {
        PaletteMix mix = gris(DryingClass.MEDIUM, Workshop.standard());
        assertThat(mixes.state(mix, NOW).workable())
                .as("le tas n'est pas etale : il prend bien plus lentement")
                .isGreaterThan(Duration.ofHours(1));
    }

    @Test
    @DisplayName("faire le menage ne retire que ce qui a pris")
    void cleaningRemovesOnlyWhatIsSpent() {
        gris(DryingClass.MEDIUM, Workshop.standard());
        int before = mixes.findAll().size();

        // A l'instant meme, rien n'a pris : le menage ne doit rien emporter.
        assertThat(mixes.forgetSpent(NOW)).isZero();
        assertThat(mixes.findAll()).hasSize(before);

        // Trois semaines plus tard, le melange n'a plus rien a faire sur la palette.
        assertThat(mixes.forgetSpent(NOW.plus(Duration.ofDays(21)))).isEqualTo(before);
        assertThat(mixes.findAll()).isEmpty();
    }
}

File: palettier-kmp/core-domain/src/commonTest/kotlin/be/asmolabs/palettier/domain/palette/PaletteMixServiceTest.kt
package be.asmolabs.palettier.domain.palette

import be.asmolabs.palettier.domain.drying.Workshop
import be.asmolabs.palettier.domain.paint.DryingClass
import be.asmolabs.palettier.domain.paint.Ventilation
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.time.Duration
import kotlin.time.Duration.Companion.days
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Instant

class PaletteMixServiceTest {

    private val now = Instant.parse("2026-03-01T10:00:00Z")
    private val service = PaletteMixService()

    private fun gris(drying: DryingClass = DryingClass.MEDIUM, workshop: Workshop = Workshop.standard()) =
        PaletteMix(
            name = "Gris rompu des ombres", hexColor = "#6B6259",
            recipe = "2 parts de terre d'ombre + 1 part de blanc",
            dryingClass = drying, mixedAt = now, workshop = workshop,
        )

    @Test
    fun `un melange frais est ouvert, et le restera quelques heures`() {
        val state = service.state(gris(), now + 30.minutes)

        assertTrue(state.isOpen)
        assertTrue(!state.isSpent)
        assertTrue(state.workable > Duration.ZERO)
    }

    @Test
    fun `passe le temps ouvert il ne se travaille plus, et plus tard il a pris`() {
        val spent = service.state(gris(DryingClass.FAST), now + 20.days)

        assertTrue(!spent.isOpen)
        assertTrue(spent.isSpent)
    }

    @Test
    fun `le temps ouvert depend de l'atelier du moment ou l'on a melange`() {
        val cold = service.state(gris(workshop = Workshop(10.0, 80.0, Ventilation.CONFINED)), now + 1.hours)
        val warm = service.state(gris(workshop = Workshop(28.0, 30.0, Ventilation.GOOD)), now + 1.hours)

        assertTrue(cold.workable > warm.workable, "un atelier froid et humide garde le melange ouvert plus longtemps")
    }

    @Test
    fun `une pate sur la palette tient plus longtemps que la meme couche sur la piece`() {
        assertTrue(service.state(gris(), now).workable > 1.hours, "le tas n'est pas etale")
    }

    @Test
    fun `le menage ne retire que ce qui a pris`() {
        val mixes = listOf(gris())

        assertEquals(emptyList(), service.spent(mixes, now))
        assertEquals(1, service.spent(mixes, now + 21.days).size)
    }
}
